python/pyspark/sql/connect/session.py

```diff
@@ -205,6 +207,31 @@ def __init__(self, connectionString: str, userId: Optional[str] = None):
         # Create the reader
         self.read = DataFrameReader(self)
 
+    def createDataFrame(self, data: "pd.DataFrame") -> "DataFrame":
```

Comment:
Actually, the implementation here isn't matched to what we have in `createDataFrame(pandas)`.

By default, the Arrow message conversion (more specifically in https://github.com/apache/spark/pull/38659/files#diff-d630cc4be6c65a3c3f7d6dbfe990f99ba992ccc26d9c3aaf6cfe46e163cb7389R514-R521) have to happen in RDD so we can parallelize this.

For a bit of history, PySpark added the initial version with RDD first, and added this local relation as an optimization for small dataset (see also https://github.com/apache/spark/pull/36683) later.

Reply:
I am fine with the current approach but the main problem here is that 1. we can't stream the input, 2. it will have the size limit (likely 4KB). cc @hvanhovell FYI

Reply:
It is impossible to match the implementation because in Pyspark to parallelize a first serialization is already happening to pass the input DF to the executors. 

In our case to even send the data to spark we have to serialize it. 

That said you're right that this currently does not support streaming of local data to the client. But the limit is not 4kb but probably whatever the max message size of GRPC is so in the megabytes. 


I think we need to add the client side streaming APIs at some point but I'd like to defer that for a bit.

Reply:
For a large `pd.dataframe`, I guess we can optimize it in this way in the future: split it into several batches, and create a `localRelation` for each batch, and finally `Union` them.